Document lobbies (#337)
discord-userdoccers/discord-userdoccers: pages/resources/lobby.mdx

      "metadata": null,
      "flags": 1
    }
  ]
}
```

### Lobby Member Object

Represents a member of a lobby.

###### Lobby Member Structure

| Field         | Type                 | Description                                                                       |
| ------------- | -------------------- | --------------------------------------------------------------------------------- |
| id            | snowflake            | The ID of the user                                                                |
| metadata?     | ?map[string, string] | The metadata of the lobby member (max 25 keys, 1024 characters per key and value) |
| flags?        | integer              | The [lobby member's flags](#lobby-member-flags)                                   |
| connected ^1^ | boolean              | Whether the member is connected to a call in lobby                                |

^1^ Only included in lobby objects returned [over the Gateway](/topics/gateway-events#lobbies).

###### Lobby Member Flags

| Value  | Name           | Description                                     |
| ------ | -------------- | ----------------------------------------------- |

> | 1 << 0 | CAN_LINK_LOBBY | Lobby member can link a text channel to a lobby |

Reviewer: ```suggestion
| 1 << 0 | CAN_LINK_LOBBY | Lobby member can link a text channel to the lobby |
```

## Endpoints

<RouteHeader method="POST" url="/lobbies">
  Create Lobby
</RouteHeader>

Creates a new lobby. Returns a [lobby](#lobby-object) object on success. Fires a [Lobby Create](/topics/gateway-events#lobby-create) Gateway event.

### JSON Params

| Field                 | Type                                               | Description                                                                                |
| --------------------- | -------------------------------------------------- | ------------------------------------------------------------------------------------------ |
| metadata?             | ?map[string, string]                               | The metadata of the lobby (max 25 keys, 1024 characters per key and value)                 |
| members?              | array[[lobby member](#lobby-member-object) object] | The users to create lobby with (max 25)                                                    |

> | idle_timeout_seconds? | integer                                            | The seconds to wait before shutting down a lobby after it becomes idle (min 5, max 604800) |

Reviewer: what's the default?

| idle_timeout_seconds? | integer                                            | How long to wait (in seconds) before shutting down a lobby after it becomes idle (min 5, max 604800, default 300) |

<RouteHeader method="PUT" url="/lobbies">
  Join or Create Lobby
</RouteHeader>

Joins an existing lobby or creates a new lobby. Returns a [lobby](#lobby-object) object on success. May fire a [Lobby Create](/topics/gateway-events#lobby-create) or [Lobby Member Add](/topics/gateway-events#lobby-member-add) Gateway event.

###### JSON Params

| Field                 | Type                 | Description                                                                                                       |
| --------------------- | -------------------- | ----------------------------------------------------------------------------------------------------------------- |
| secret                | string               | The lobby secret (max 250 characters)                                                                             |
| lobby_metadata        | ?map[string, string] | The metadata of the lobby (max 25 keys, 1024 characters per key and value)                                        |

> | member_metadata       | ?map[string, string] | The metadata of the lobby member (max 25 keys, 1024 characters per key and value)                                 |

Reviewer: this is not true; if they are both nullable and optional, mark them as such

<Alert type="warn">

This endpoint is not usable by user accounts.

</Alert>

###### JSON Params

| Field                 | Type                                               | Description                                                                                                       |
| --------------------- | -------------------------------------------------- | ----------------------------------------------------------------------------------------------------------------- |
| metadata?             | ?map[string, string]                               | The metadata of the lobby (max 25 keys, 1024 characters per key and value)                                        |
| members?              | array[[lobby member](#lobby-member-object) object] | The members of the lobby (max 25)                                                                                 |
| idle_timeout_seconds? | integer                                            | How long to wait (in seconds) before shutting down a lobby after it becomes idle (min 5, max 604800, default 300) |

<RouteHeader method="PUT" url="/lobbies">
  Join or Create Lobby

> </RouteHeader>

Reviewer: this (and others) are missing the supportsOAuth2 directive
also if the ep only works w oauth2 (and not with a user token) it needs a warning for that too, like the game relationship endpoints have

# Lobbies

### Lobby Object

Represents a lobby within Discord.

###### Lobby Structure

| Field           | Type                                                | Description                                       |
| --------------- | --------------------------------------------------- | ------------------------------------------------- |
| id              | snowflake                                           | The ID of the lobby                               |
| application_id  | snowflake                                           | The ID of the application that created the lobby  |

> | metadata        | ?map[string, string]                                | The metadata of the lobby (max total length 1000) |

Reviewer: What's the restriction on each key?

| members         | array[[lobby member](#lobby-member-object) object]  | The members of the lobby                          |
| linked_channel? | [channel](/resources/channel#channel-object) object | The guild channel linked to the lobby             |

###### Example Lobby Object

```json
{
  "id": "1350184413060661332",
  "application_id": "891436243903728565",
  "metadata": {
    "topic": "balls"
  },
  "members": [
    {
      "id": "852892297661906993",
      "metadata": null,
      "flags": 1
    }
  ]
}
```

### Lobby Member Object

Represents a member of a lobby.

###### Lobby Member Structure

| Field             | Type                 | Description                                              |
| ----------------- | -------------------- | -------------------------------------------------------- |
| id                | snowflake            | The ID of the user                                       |

> | metadata?         | ?map[string, string] | The metadata of the lobby member (max total length 1000) |

Reviewer: same here

| flags?            | integer              | The [lobby member's flags](#lobby-member-flags)          |

> | connected ^1^ ^2^ | boolean              | Whether the member is connected to call in a lobby       |

Reviewer: ```suggestion
| connected ^1^ ^2^ | boolean              | Whether the member is connected to a call in the lobby       |
```

> ^1^ This field can be only received.
> 
> ^2^ This field is provided only if object was received over the Gateway.

Reviewer: ```suggestion
^1^ Only included in guild objects returned [over the Gateway](/topics/gateway-events#lobbies).
```
these are redundant

###### Lobby Member Flags

| Value  | Name           | Description                                     |
| ------ | -------------- | ----------------------------------------------- |
| 1 << 0 | CAN_LINK_LOBBY | Lobby member can link a text channel to a lobby |

### Lobby Message Object

A message sent in a lobby within Discord.

> {/* TODO: Should this be documented as superset of a normal Message object? */}

Reviewer: yes definitely. should also have a callout for the additional fields in ephemeral dm messages (recipient_id, recipients, channel (and the special channel object), all in the partial message structure

| Field                 | Type                 | Description                                                                                |
| --------------------- | -------------------- | ------------------------------------------------------------------------------------------ |
| secret                | string               | The lobby secret                                                                           |
| lobby_metadata        | ?map[string, string] | The metadata of the lobby (max total length 1000)                                          |
| member_metadata       | ?map[string, string] | The metadata of the lobby member (max total length 1000)                                   |
| idle_timeout_seconds? | integer              | The seconds to wait before shutting down a lobby after it becomes idle (min 5, max 604800) |

<RouteHeader method="GET" url="/lobbies/{lobby.id}">
  Get Lobby
</RouteHeader>

Returns a [lobby](#lobby-object) object for the given lobby ID. User must be a member of the lobby.

<Alert type="warn">

This endpoint is not usable by user accounts.

</Alert>

<RouteHeader method="PATCH" url="/lobbies/{lobby.id}">
  Modify Lobby
</RouteHeader>

> Modifies a lobby's settings. Application must be creator of the lobby. Returns the updated [lobby](#lobby-object) object on success. Fires a [Lobby Update](/topics/gateway-events#lobby-update) Gateway event.

Reviewer: ```suggestion
Modifies a lobby's settings. Application must be the creator of the lobby. Returns the updated [lobby](#lobby-object) object on success. Fires a [Lobby Update](/topics/gateway-events#lobby-update) Gateway event.
```

same change in the rest

| members?              | array[[lobby member](#lobby-member-object) object] | The users to create lobby with (max 25)                                                    |
| idle_timeout_seconds? | integer                                            | The seconds to wait before shutting down a lobby after it becomes idle (min 5, max 604800) |

<RouteHeader method="DELETE" url="/lobbies/{lobby.id}">
  Delete Lobby
</RouteHeader>

Deletes a lobby. Application must be creator of the lobby. Returns a 204 empty response on success. Fires a [Lobby Delete](/topics/gateway-events#lobby-delete) Gateway event.

<Alert type="warn">

This endpoint is not usable by user accounts.

</Alert>

<RouteHeader method="PUT" url="/lobbies/{lobby.id}/members/{user.id}">
  Add Lobby Member
</RouteHeader>

Adds a member to the lobby. Returns the [lobby member](#lobby-member-object) object. Fires a [Lobby Member Add](/topics/gateway-events#lobby-member-add) or [Lobby Member Update](/topics/gateway-events#lobby-member-update) Gateway event.

<Alert type="note">

> If user is already a member of the lobby, this will update provided fields and fire [Lobby Member Update](/topics/gateway-events#lobby-member-update) Gateway event instead.

Reviewer: ```suggestion
If the user is already a member of the lobby, this will update the provided fields and fire a [Lobby Member Update](/topics/gateway-events#lobby-member-update) Gateway event instead.
```

so many missing articles xd
also, does this *update* or *replace* the provided fields (e.g. will omitting one reset it to default?)

</Alert>

### JSON Params

| Field     | Type                 | Description                                              |
| --------- | -------------------- | -------------------------------------------------------- |
| metadata? | ?map[string, string] | The metadata of the lobby member (max total length 1000) |
| flags?    | integer              | The [lobby member's flags](#lobby-member-flags)          |

<RouteHeader method="DELETE" url="/lobbies/{lobby.id}/members/{user.id}">
  Remove Lobby Member
</RouteHeader>

Removes a member from the lobby. Returns a 204 empty response on success. Fires a [Lobby Member Remove](/topics/gateway-events#lobby-member-remove) Gateway event.

> {/* TODO: do we clarify whether this returns 204 even if provided user is not member of lobby? */}

Reviewer: no this is fairly standard

<RouteHeader method="DELETE" url="/lobbies/{lobby.id}/members/@me" supportsOAuth2="lobbies.write">
  Leave Lobby
</RouteHeader>

Removes the current user from the lobby. Returns a 204 empty response on success. Fires a [Lobby Delete](/topics/gateway-events#lobby-delete) and [Lobby Member Remove](/topics/gateway-events#lobby-member-remove) Gateway event.

{/* TODO: do we clarify whether this returns 204 even if provided user is not member of lobby? */}

<RouteHeader method="PATCH" url="/lobbies/{lobby.id}/channel-linking" supportsOAuth2="lobbies.write">
  Set Linked Channel
</RouteHeader>

Links or unlinks a channel to the lobby. Returns a [lobby](#lobby-object) object on success. Fires a [Lobby Update](/topics/gateway-events#lobby-update) and [Channel Update](/topics/gateway-events#channel-update) Gateway event.

> The application must be a creator of the lobby, or user must have `CAN_LINK_LOBBY` flag.

Reviewer: ```suggestion
The application must be the creator of the lobby or the member must have the [`CAN_LINK_LOBBY` flag](#lobby-member-flags).
```

### JSON params

| Field       | Type      | Description                                |
| ----------- | --------- | ------------------------------------------ |
| channel_id? | snowflake | The ID of the channel to link to the lobby |

<RouteHeader method="POST" url="/lobbies/{lobby.id}/messages" supportsOAuth2="lobbies.write">
  Create Lobby Message
</RouteHeader>

> Posts a message to a text-based channel. Returns a partial [lobby message](#partial-lobby-message-object) object on success. Fires a [Lobby Message Create](/topics/gateway-events#lobby-message-create) and [Message Create](/topics/gateway-events#message-create) Gateway event.

Reviewer: ```suggestion
Posts a message to a lobby. Returns a partial [lobby message](#partial-lobby-message-object) object on success. Fires a [Lobby Message Create](/topics/gateway-events#lobby-message-create) and [Message Create](/topics/gateway-events#message-create) Gateway event.
```
